code/modules/mob/living/carbon/human/ai/firearm_appraisal.dm

```diff
@@ -12,55 +12,57 @@ GLOBAL_LIST_INIT_TYPED(firearm_appraisals, /datum/firearm_appraisal, build_firea
 	var/optimal_range = 6
 	/// Maximum engagement range, stop firing at this distance
-	var/maximum_range = 9
+	var/maximum_range = 16
 	/// How many rounds to fire in 1 burst at most
 	var/burst_amount_max = 8
 	/// List of types that set the human AI to this appraisal type
 	var/list/gun_types = list()
 	/// If TRUE, this gun is disposable and isn't worth trying to reload
 	var/disposable = FALSE
 
-/// List of things we do before our next fire based on weapon type
+/// List of things we do before beginning to spray bullets based off weapon type
 /datum/firearm_appraisal/proc/before_fire(obj/item/weapon/gun/firearm, mob/living/carbon/user, datum/human_ai_brain/AI)
-	SHOULD_CALL_PARENT(TRUE) // Every weapon can be twohanded
+	SHOULD_CALL_PARENT(TRUE) // Every weapon may be twohanded or have safety
+	set waitfor = FALSE
 
 	AI.ensure_primary_hand(firearm)
 	if((firearm.flags_item & TWOHANDED) && !(firearm.flags_item & WIELDED))
-		firearm.wield(user)
-		sleep(max(firearm.wield_delay, AI.short_action_delay * AI.action_delay_mult))
+		AI.wield_primary_sleep()
 
 	if(firearm.flags_gun_features & GUN_TRIGGER_SAFETY)
 		firearm.flags_gun_features ^= GUN_TRIGGER_SAFETY
 		firearm.gun_safety_handle(user)
 
-/datum/firearm_appraisal/proc/do_reload(obj/item/ammo_magazine/mag, mob/living/carbon/user, datum/human_ai_brain/AI)
+/// Reload sequence per weapon type, override as needed
+/datum/firearm_appraisal/proc/do_reload(obj/item/weapon/gun/firearm, obj/item/ammo_magazine/mag, mob/living/carbon/user, datum/human_ai_brain/AI)
 	AI.unholster_primary()
-	AI.ensure_primary_hand(AI.primary_weapon)
-	AI.primary_weapon.unwield(user)
+	AI.ensure_primary_hand(firearm)
+	firearm.unwield(user)
 	sleep(AI.short_action_delay * AI.action_delay_mult)
-	if(!(AI.primary_weapon?.flags_gun_features & GUN_INTERNAL_MAG) && AI.primary_weapon?.current_mag)
-		AI.primary_weapon?.unload(user, FALSE, TRUE, FALSE)
+	if(!(firearm?.flags_gun_features & GUN_INTERNAL_MAG) && firearm?.current_mag)
+		firearm?.unload(user, FALSE, TRUE, FALSE)
 	user.swap_hand()
 	sleep(AI.micro_action_delay * AI.action_delay_mult)
 	AI.equip_item_from_equipment_map(HUMAN_AI_AMMUNITION, mag)
 	sleep(AI.short_action_delay * AI.action_delay_mult)
 	if(istype(mag, /obj/item/ammo_magazine/handful))
 		for(var/i in 1 to mag.current_rounds)
-			AI.primary_weapon?.attackby(mag, user)
+			firearm?.attackby(mag, user)
 			sleep(AI.micro_action_delay * AI.action_delay_mult)
 		if(!QDELETED(mag) && (mag.current_rounds > 0))
 			var/storage_slot = AI.storage_has_room(mag)
 			if(storage_slot)
-				AI.store_item(mag, storage_slot)
+				AI.store_item(mag, storage_slot, HUMAN_AI_AMMUNITION)
 			else
 				user.drop_held_item(mag)
 	else
-		AI.primary_weapon?.attackby(mag, user)
+		firearm?.attackby(mag, user)
 	sleep(AI.short_action_delay * AI.action_delay_mult)
 	user.swap_hand()
-	AI.primary_weapon?.wield(user)
+	AI.wield_primary_sleep()
 
 /datum/firearm_appraisal/sniper
+	optimal_range = 7
```

Comment:
why

Reply:
Parent optimal range is 6, if you have a sniper rifle it would be logical to try and back off while shooting point-blank, at least until the screen edge